bp_common/src/include/bp_common_cache_service.vh

```diff
@@ -0,0 +1,148 @@
+
+ // read tag mem packets for writeback and transfer (Used for UCE)
+ e_cache_tag_mem_read
+
+} bp_cache_tag_mem_opcode_e;
+
+`define bp_cache_tag_mem_opcode_width $bits(bp_cache_tag_mem_opcode_e)
+
+// Stat mem pkt opcodes
+typedef enum logic [1:0] {
+ // clear all dirty bits and LRU bits to zero for given index.
+ e_cache_stat_mem_set_clear,
+
+ // read stat_info for given index.
+ e_cache_stat_mem_read,
+
+ // clear dirty bit for given index and way_id.
+ e_cache_stat_mem_clear_dirty
+
+} bp_cache_stat_mem_opcode_e;
+
+`define bp_cache_stat_mem_opcode_width $bits(bp_cache_stat_mem_opcode_e)
+
+// data mem pkt structure
+`define declare_bp_cache_data_mem_pkt_s(sets_mp, ways_mp, data_width_mp)                \
+  typedef struct packed                                                                 \
+  {                                                                                     \
+    logic [`BSG_SAFE_CLOG2(sets_mp)-1:0]      index;                                    \
+    logic [`BSG_SAFE_CLOG2(ways_mp)-1:0]      way_id;                                   \
+    logic [data_width_mp-1:0]                 data;                                     \
+    bp_cache_data_mem_opcode_e                opcode;                                   \
+  }  bp_cache_data_mem_pkt_s;
+
+`define bp_cache_data_mem_pkt_width(sets_mp, ways_mp, data_width_mp) \
+  (`BSG_SAFE_CLOG2(sets_mp)+`BSG_SAFE_CLOG2(ways_mp)+data_width_mp \
+   +`bp_cache_data_mem_opcode_width)
+
+// tag mem pkt structure
+`define declare_bp_cache_tag_mem_pkt_s(sets_mp, ways_mp, tag_width_mp)           \
+  typedef struct packed {                                                        \
+    logic [`BSG_SAFE_CLOG2(sets_mp)-1:0]        index;                           \
+    logic [`BSG_SAFE_CLOG2(ways_mp)-1:0]        way_id;                          \
+    logic [`bp_coh_bits-1:0]                    state;                           \
+    logic [tag_width_mp-1:0]                    tag;                             \
+    bp_cache_tag_mem_opcode_e                   opcode;                          \
+  }  bp_cache_tag_mem_pkt_s;
+
+`define bp_cache_tag_mem_pkt_width(sets_mp, ways_mp, tag_width_mp) \
+  (`BSG_SAFE_CLOG2(sets_mp)+`BSG_SAFE_CLOG2(ways_mp)+`bp_coh_bits+tag_width_mp+`bp_cache_tag_mem_opcode_width)
+
+`define declare_bp_cache_stat_mem_pkt_s(sets_mp, ways_mp)                 \
+  typedef struct packed {                                                 \
+    logic [`BSG_SAFE_CLOG2(sets_mp)-1:0]    index;                        \
+    logic [`BSG_SAFE_CLOG2(ways_mp)-1:0]    way_id;                       \
+    bp_cache_stat_mem_opcode_e              opcode;                       \
+  } bp_cache_stat_mem_pkt_s;
+
+`define bp_cache_stat_mem_pkt_width(sets_mp, ways_mp) \
+  (`BSG_SAFE_CLOG2(sets_mp)+`BSG_SAFE_CLOG2(ways_mp)+`bp_cache_stat_mem_opcode_width)
+
+`define declare_bp_cache_if_widths(ways_mp, sets_mp, tag_width_mp, lce_data_width_mp)                                \
```

Comment:
Include declare_bp_cache_req_widths here

Reply:
If you want cache_req_widths to create localparams for the data, tag and stat men packets, then we are calling it cache_if_widths. cache_req_widths is associated with cache_req_o

Reply:
If you want cache_req_widths to create localparams for the data, tag and stat men packets, then we are calling it cache_if_widths. cache_req_widths is associated with cache_req_o

Reply:
Sorry, I think I wasn't clear.  I meant

`declare_bp_cache_if_widths
, localparam bp_cache_req_width_lp = `bp_cache_req_width(data_width_mp, ways_mp, paddr_width_mp)
, localparam bp_cache_data_mem_pkt_width_lp = 
....
`

Reply:
Additionally, drop the bp_ prefix for the width names to follow convention